### ISSUE TITLE
The lifetime of `&self` in `Container::exec` shouldn't be tied to `'docker`

### ISSUE THREAD

**Issue by azdle** [vv9k/docker-api-rs#59]
- Crate version: `master` @ [6c0a8f4](https://github.com/vv9k/docker-api-rs/commit/6c0a8f4bc403d24434f1a9774e4d09d957522640)

I believe that [`Container::exec`](https://github.com/vv9k/docker-api-rs/blob/6c0a8f4bc403d24434f1a9774e4d09d957522640/src/api/container.rs#L218-L224) shouldn't tie the lifetime of `&self` to `'docker`, but rather `Container` should have a generic lifetime that would let you tie the `'docker` lifetime to `self.docker` itself.

From looking at the implementation, I think the current way isn't required and probably wasn't the intent based on the name of the lifetime.

(Though, I am getting a bit lost in all the macros that actually generate the types. I may be missing a reason why it has to be this way.)

---

I'm trying to use `Container::exec` within a function in my library and I want to return the `Multiplexer` it gives me. However, the lifetime of the `Multiplexer` is getting tied to the `Container` struct.

A simplified example:

```
pub async fn shell(&self, container_id: &str) -> Result<Multiplexer> {
      trace!(container_id, "attach to container");

      let mut container = self.client.containers().get(container_id);

      let create_opts = ExecCreateOpts::builder()
          .attach_stdout(true)
          .attach_stderr(true)
          .attach_stdin(true)
          .detach_keys("ctrl-d")
          .tty(true)
          .command(["/bin/sh"])
          .build();
      let start_opts = ExecStartOpts::builder().tty(true).build();
      let io = container.exec(&create_opts, &start_opts).await?;

      Ok(io)
}
```

gives

```
error[E0515]: cannot return value referencing local variable `container`
   --> conductor/src/containers/mod.rs:819:17
    |
817 |                 let io = container.exec(&create_opts, &start_opts).await?;
    |                          ----------------------------------------- `container` is borrowed here
818 |
819 |                 Ok(io)
    |                 ^^^^^^ returns a value referencing data owned by the current function
```

**Comment by azdle** [comment]
Oh, crikey, I now see why it's this way. `Container` has an owned clone of `Docker` so the `Multiplexer` is actually holding a reference to something in `Container`. :/

What I suggested above would only make sense if `Container` was changed to just hold a reference to `Docker`.

A probably better solution would be for `Container::exec` to clone the `Docker` too and have that get stored in the `Multiplexer`. I think that's probably more in line with how you're handing sharing of the underlying Hyper connections. Is that possible?

Or, is there something with the current API I'm missing that would allow me to do this that I'm just missing?

**Comment by vv9k** [comment]
That's a valid observation. We moved to using owned clone of `Docker` as from what I've read in hyper documentation it is cheap to clone the client and it's a good way to avoid lifetime issues. The exec part was just somehow omitted during this change.

I've created an initial PR [here](https://github.com/vv9k/docker-api-rs/pull/60). Let me know if it resolves your issues, I've run the test suite and everything seems to work as expected.
